source/retryable-writes/retryable-writes.rst

```diff
@@ -511,6 +513,15 @@ experience across all drivers.
 Why expose users to server-side errors for unsupported operations?
 ------------------------------------------------------------------
 
+.. todo:: Determine how this should change. Do we end up using whitelists for
+   supported operations (based on wire protocol verison) and avoid injecting
+   transaction IDs on all write ops, or rely on the server to raise an error if
+   unsupported ops would be retried? In both cases, it's possible for users to
+   be unaware that some operations are not supported until some error in prod.
+   Given the MongoClient-wide scope of retryWrites, we likely don't want to
+   prohibit unsupported operations outright (either in the client API or by
+   having the server raise an exception on the first attempt).
+
```

Comment:
I concur with the objections to whitelists.

@ajdavis makes a good point that `{ w: 0 }` will need to be permitted in the interest of application portability.

Having the server ignore the transaction ID (sessionID/txnNum pair) on first attempts and erroring on the second attempt still has the caveat of users not realizing that some operations cannot be retried during testing. I know that we settled on only soliciting `retryWrites` as a client-level option (https://github.com/mongodb/specifications/pull/174/files#r135094282), but allowing it on the database and collection objects could help address the "didn't realize until production" situation if that's still a concern. I believe when we last discussed not supporting the option on database and collection objects, it was to avoid creating a dependency on the ThreadLocal driver session object; however, since the drivers now fetch a server session from the pool as needed per operation, ThreadLocal bundling is no longer an issue.

Is this something we'd like to discuss further, or should we just formally request that the server changes its behavior to:

> The server should only return an error if a retry is actually attempted and not supported.